Integer overflow with $GENERATE test on armhf

Hi,

I'm troubleshooting a test error we started to see in armhf:
```
=== RUN   TestGenerateRangeGuard
    TestGenerateRangeGuard: generate_test.go:81: expected

        @ IN SOA ns.test. hostmaster.test. ( 1 8h 2h 7d 1d )
        $GENERATE 0-2 dhcp-${2147483647,4,d} A 10.0.0.$

        to fail, but got no error
--- FAIL: TestGenerateRangeGuard (0.01s)
```
I believe this was introduced in commit c0283a2028317d3fb6574f9a06e68443cf6b89cd which added this check:
```golang
                        if r.start+offset < 0 || r.end+offset > 1<<31-1 {
                                return 0, r.parseError("bad offset in $GENERATE", si+3+sep)
                        }
```
The test will reach (1<<31-1)+2, meaning `r.end+offset` on 32bits systems will become negative and thus not trigger the condition above.

> Edit: That won't work for negative offsets. The better solution would probably be to replace the check with: `int64(end) + int64(offset) > 1<<31-1`.

That seems to work for this test.

I see another failure though:

```
=== RUN   TestCrasherString
=== RUN   TestCrasherString/$GENERATE_0-300103"$$GENERATE_2-2
=== RUN   TestCrasherString/$GENERATE_0-5414137360
    TestCrasherString/$GENERATE_0-5414137360: generate_test.go:230: Expecting error bad range in $GENERATE, got dns: bad stop in $GENERATE range: "0-5414137360"
 at line: 1:22
=== RUN   TestCrasherString/$GENERATE_______11522-3668518066406258
    TestCrasherString/$GENERATE_______11522-3668518066406258: generate_test.go:230: Expecting error bad range in $GENERATE, got dns: bad stop in $GENERATE range
: "11522-3668518066406258" at line: 1:38
=== RUN   TestCrasherString/$GENERATE_0-200"(;00000000000000_$$GENERATE_0-0
=== RUN   TestCrasherString/$GENERATE_6-2048_$$GENERATE_6-036160_$$$$ORIGIN_\$
--- FAIL: TestCrasherString (0.01s)
    --- PASS: TestCrasherString/$GENERATE_0-300103"$$GENERATE_2-2 (0.00s)
    --- FAIL: TestCrasherString/$GENERATE_0-5414137360 (0.00s)
    --- FAIL: TestCrasherString/$GENERATE_______11522-3668518066406258 (0.00s)
    --- PASS: TestCrasherString/$GENERATE_0-200"(;00000000000000_$$GENERATE_0-0 (0.00s)
    --- PASS: TestCrasherString/$GENERATE_6-2048_$$GENERATE_6-036160_$$$$ORIGIN_\$ (0.00s)
```

--- a/generate.go
+++ b/generate.go
@@ -188,7 +188,7 @@ func (r *generateReader) ReadByte() (byte, error) {
 			if errMsg != "" {
 				return 0, r.parseError(errMsg, si+3+sep)
 			}
-			if r.start+offset < 0 || r.end+offset > 1<<31-1 {
+			if r.start+offset < 0 || int64(r.end) + int64(offset) > 1<<31-1 {
 				return 0, r.parseError("bad offset in $GENERATE", si+3+sep)
 			}
 

--- a/generate.go
+++ b/generate.go
@@ -20,13 +20,13 @@ import (
 // of $ after that are interpreted.
 func (zp *ZoneParser) generate(l lex) (RR, bool) {
 	token := l.token
-	step := 1
+	step := int64(1)
 	if i := strings.IndexByte(token, '/'); i >= 0 {
 		if i+1 == len(token) {
 			return zp.setParseError("bad step in $GENERATE range", l)
 		}
 
-		s, err := strconv.Atoi(token[i+1:])
+		s, err := strconv.ParseInt(token[i+1:], 10, 64)
 		if err != nil || s <= 0 {
 			return zp.setParseError("bad step in $GENERATE range", l)
 		}
@@ -40,12 +40,12 @@ func (zp *ZoneParser) generate(l lex) (RR, bool) {
 		return zp.setParseError("bad start-stop in $GENERATE range", l)
 	}
 
-	start, err := strconv.Atoi(sx[0])
+	start, err := strconv.ParseInt(sx[0], 10, 64)
 	if err != nil {
 		return zp.setParseError("bad start in $GENERATE range", l)
 	}
 
-	end, err := strconv.Atoi(sx[1])
+	end, err := strconv.ParseInt(sx[1], 10, 64)
 	if err != nil {
 		return zp.setParseError("bad stop in $GENERATE range", l)
 	}
@@ -75,10 +75,10 @@ func (zp *ZoneParser) generate(l lex) (RR, bool) {
 	r := &generateReader{
 		s: s,
 
-		cur:   start,
-		start: start,
-		end:   end,
-		step:  step,
+		cur:   int(start),
+		start: int(start),
+		end:   int(end),
+		step:  int(step),
 
 		file: zp.file,
 		lex:  &l,
@@ -229,19 +229,19 @@ func modToPrintf(s string) (string, int, string) {
 		return "", 0, "bad base in $GENERATE"
 	}
 
-	offset, err := strconv.Atoi(offStr)
+	offset, err := strconv.ParseInt(offStr, 10, 64)
 	if err != nil {
 		return "", 0, "bad offset in $GENERATE"
 	}
 
-	width, err := strconv.Atoi(widthStr)
+	width, err := strconv.ParseInt(widthStr, 10, 64)
 	if err != nil || width < 0 || width > 255 {
 		return "", 0, "bad width in $GENERATE"
 	}
 
 	if width == 0 {
-		return "%" + base, offset, ""
+		return "%" + base, int(offset), ""
 	}
 
-	return "%0" + widthStr + base, offset, ""
+	return "%0" + widthStr + base, int(offset), ""
 }

